xarray/core/datatree.py

```diff
@@ -169,33 +166,32 @@ def update(self, other) -> NoReturn:
         )
 
     # FIXME https://github.com/python/mypy/issues/7328
-    @overload
-    def __getitem__(self, key: Mapping) -> Dataset:  # type: ignore[misc]
+    @overload  # type: ignore[override]
```

Comment:
There's a few new `type: ignore[override]` annotations on `DatasetView` methods. @flamingbear and I spent a while today trying to disentangle things here.

Essentially, the issue comes down to `DatasetView` inheriting from `Dataset`, which uses `typing.Self` as a return type in a few places. This would mean `DatasetView` would have a return type of `DatasetView`. But, instead, these return types are explicitly a `Dataset`. ([mypy has some examples](https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides) mentioning that subclasses shouldn't have more generalised return types than their parents, which seems related - thanks @flamingbear for the reference)

I'm not sure that just ignoring the error is the best thing to do, but I don't think we had a better idea for an implementation. I think @TomNicholas has mentioned a future step of reworking the class inheritance, but I'm not sure if that would also cover this.

Reply:
Hmm. So I guess I have violated the Liskov substitution principle here... Does that represent an actual possible failure mode? Like are there any places where "if you can use a `Dataset` you can use a `DatasetView`" fails? I suppose passing a `DatasetView` into any function that attempts to call `__setitem__` would violate this.

Are there alternative designs that would give us immutability without `DatasetView` being a subclass of `Dataset`? One alternative might be something like

```python
class FrozenDataset:
    _dataset: Dataset

    def __init__(self, ds):
        self._dataset = ds

    def __getattr__(self, name):
        # Forward the method call to the Dataset class
        if name in ['__setitem__', ...]:
            raise AttributeError("Mutation of the DatasetView is not allowed, ...")
        else:
            return getattr(self._ds, name)
```
but then that would not pass an `isinstance` check, i.e. `isinstance(dt.ds, Dataset)` would return `False`.

> @TomNicholas has mentioned a future step of reworking the class inheritance, but I'm not sure if that would also cover this.

I was just talking about having both `Dataset` and `DataTree` inherit from a common `DataMapping` class that holds variables. But I don't think that would cover this, as that `DataMapping` should also be returning `Self`.

Reply:
This is much more like the `Frozen` class that xarray sometimes uses.

Reply:
So looking at this in Owen's absence. It seems like there are two paths. 

  + Use a frozenDataset type thing, that fails the isinstance check?
  + Overriding the mypy errors understanding the the DatasetView is an library internal implementation and we know not to misuse it?
  
The first seems like the right thing to do, but I don't know what breaks with the isinstance check or why that would be necessary.
Is there some ABC that both Dataset and DatasetView could implement to pass the isinstance check and is that a big change? Also, this may show that I'm not afraid to ask the dumb questions in public.

Reply:
I'm going to make it a wrapper.

Reply:
Coming back to this with things that have been tried (and sadly not succeeded to date):

* First up, I tried to implement the suggested `FrozenDataset` wrapper. There was an issue with using `getattr` - as it can't be used to intercept magic methods on Python classes.
* Next: Trying to implement a `Metaclass` to allow interception of magic methods (initially inspired by [this snippet](https://stackoverflow.com/a/9059858)). This proved tricky to do (I didn't quite get it fully working) and felt very much like it was an overly complicated solution for the problem we were trying to solve.
* Next: Trying a mix-in to overwrite the affected methods. I put a time-box on this attempt as we want to unblock the migration. I did not get this working in the allocated time.
* Lastly: Conceding defeat and adding the `type ignore` statements to cover this.

Not ideal, but the `DatasetView` and `Datatree.ds` property both have usage over the last couple of years without significant issues. I opened #8855 to capture that we need to work on a better fix at a later date.